homeassistant/components/mqtt/client.py

```diff
@@ -657,36 +656,31 @@ async def _async_perform_subscriptions(self) -> None:
         # Since we do not know if a published value is retained we need to
         # (re)subscribe, to ensure retained messages are replayed
 
-        def _process_client_subscriptions() -> list[tuple[int, int]]:
+        if not self._pending_subscriptions:
+            return
+
+        subscriptions: dict[str, int] = self._pending_subscriptions
+        self._pending_subscriptions = {}
+
+        def _process_client_subscriptions() -> tuple[int, int]:
             """Initiate all subscriptions on the MQTT client and return the results."""
-            subscribe_result_list = []
+            subscribe_result_list: list[tuple[int, int]] = []
+            result, mid = self._mqttc.subscribe(list(subscriptions.items()))
+            subscribe_result_list.append((result, mid))
             for topic, qos in subscriptions.items():
-                result, mid = self._mqttc.subscribe(topic, qos)
-                subscribe_result_list.append((result, mid))
                 _LOGGER.debug("Subscribing to %s, mid: %s, qos: %s", topic, mid, qos)
```

Comment:
Move this print (and the for loop above it) to the existing `if result == 0:` block before we do the `await self._wait_for_mid(mid)`.

We should reduce the time with the lock held as much as possible, and there is no reason to keep the lock held while printing debug information (which could take a while due to disk I/O).

Reply:
For readability of the logs